Lowpriority format (#218)
northeastern-rc/rc-public-documentation: docs/source/hardware/low_priority_faqs.md

(low-priority-faqs)=

> # Low Priority Partition FAQs

Reviewer: That is a good question and something we can look into. Thank you!!

## Partition Information

> ::::{dropdown} 1. What is the `lowpriority` partition and why is it important?

Reviewer: Not sure on the difficulty on change the color of that. It sounds like it would be a CSS change. We can check on that. If it isn't something easily changeable, we could drop the ` ` around the lowpriority in the title. Thank you!!

> `lowpriority` is a new partition on Discovery that allows the research community to use **private resources** on the cluster when they are idle. This new partition has hardware not otherwise available to the general research community and, in time, could double the resources available to NURC users. This is a common practice in HPC clusters to optimize the use of idle private resources that consume power and cooling. 

Reviewer: addressed and pushed. Thank you!!

More information about the low priority partition and examples on how to use it can be 
found {ref}`low-priority`.
::::

::::{dropdown} 2. What is a "low priority" job?
A job submitted to the `lowpriority` partition that comes from a user **who is NOT** a member of the private partition on which the job is scheduled to run.

A "low priority" job will be terminated from the `lowpriority` partition and re-queued if a member of the private partition submits a job to the private partition.
::::

::::{dropdown} 3. What is a “high priority” job?

> A job submitted to a private partition by a user **who is** a member of that private partition, on which a low priority job is running.

Reviewer: I agree, the structure is strange and high priority job would just be a job that is submitted to the partition by a member of that group. I think it was just said that way to introduce there is now two levels of jobs. Also all questions are good! 

I made the change and pushed it. Thank you!!

::::

> ::::{dropdown} 4. How does SLURM handle low priority jobs?

Reviewer: Addressed and pushed. Thank you!!

> In SLURM, jobs are submitted to partitions. SLURM is configured such that jobs submitted to the `lowpriority`partition are considered for execution on private partitions only when the requested job resources are idle.

Reviewer: Made the change and pushed it. Thank you!!

In other words, a low priority job will be executed only if the private partition has resources to execute a job submitted to it. However, if a member of the private partition submits a job during this time, then the low priority job will be automatically terminated and re-queued (preempted). See {reF}`low-priority-preemption`.

> Jobs submitted to the `lowpriority` partition have the lowest priority of any class of jobs and are only considered after all other high priority jobs have been placed. Jobs submitted by private partition members to their own partition always have higher priority.

Reviewer: Addressed and pushed. Thank you!!

::::

::::{dropdown} 5. Is the `lowpriority` partition default?

> No, the default partition will still be the `short` partition. This ensures that by default, jobs are not subject to preemption.

Reviewer: Change made and pushed. Thank you!!

::::

::::{dropdown} 6. How do I submit a job to the `lowpriority` partition?
The `lowpriority` partition is a distinct partition that must be requested explicitly. SLURM accepts a comma-separated list of partitions. If a job is scheduled to run in the `lowpriority` partition, it may be preempted and re-queued. This does not happen in other partitions. The `lowpriority` partition is the only partition that can be preempted. See examples in using {ref}`low-priority`.
::::

::::{dropdown} 7. How can I ensure that my `lowpriority` partition job can run on nodes with sufficient resources?
It is important to specify the required resources in the job script, including memory, CPUs and GPUs. For CPUs, this can include the number of cores and the micro-architecture and for GPUs this is typically handled by “class”. The GPU classes that are configured on the HPC system include P100, V100, T4, and A100. See our documentation on {ref}`slurm-guide-index` and {ref}`working-gpus` for additional information.
::::

::::{dropdown} 8. What is the downside of submitting a job to the `lowpriority` partition?

> If a job is submitted to the `lowpriority` partition and a high priority job comes through that requires resources currently occupied by the low priority job, then that low priority job will be stopped/suspended within 30s and re-queued. 

Reviewer: Changes made and pushed. Thank you!!

::::

::::{dropdown} 9. When should I NOT use the `lowpriority` partition to run my jobs?

> Jobs running on the `lowpriority` partition always carry the risk of being suspended before their wall time ends if a high priority job requests those resources while the low priority job is running. If you have checkpointing (see {ref}`checkpoint-jobs`) implemented in your workflow, such abrupt suspension of jobs would not be an issue, since your intermediate calculations/data are saved, and you can re-start your jobs from the point of their suspension. However, if you do not have checkpointing techniques implemented, and/or you expect your jobs to run for a while (and re-running them in the event of preemption means it is going to start all over again and take even longer to complete), it is not ideal to use the `lowpriority` partition. 

Reviewer 2: ```suggestion
Jobs running on the `lowpriority` partition always carry the risk of being suspended by a high priority job before their wall time ends. If you have checkpointing (see {ref}`checkpoint-jobs`) implemented in your workflow, such abrupt suspension of jobs would not be an issue, since your intermediate calculations/data are saved, and you can re-start your jobs from the point of their suspension. However, if you do not have checkpointing techniques implemented, and/or you expect your jobs will run for a long time, it is not ideal to use the `lowpriority` partition. 
```

Change link to: 'checkpointing implemented in your workflow'.

::::

:::{button-link} faq.html
:color: primary
:align: center
Back to the top
:::
## Resources

::::{dropdown} 1. What are the time limits on the `lowpriority` partition?
Partition limits for the `lowpriority` partition are the same as the default partition (`short`), which has a 4h default walltime and a 24h max-walltime.
::::

::::{dropdown} 2. What are the resource and job limits on the `lowpriority` partition?

> Currently, the resource limits and the number for CPU-supported low priority jobs on the `lowpriority` partition are the same as the default partition (see {ref}`partition-names`). The resource limits and the number for GPU-supported low priority jobs on the `lowpriority` partition are the same as the gpu partition (see {ref}`partition-names`).

Reviewer 2: ```suggestion
Currently, the resource limits and the number for CPU-supported low priority jobs on the `lowpriority` partition are the same as the default partition (see {ref}`partition-names`). The resource limits and the number for GPU-supported low priority jobs on the `lowpriority` partition are the same as the `gpu` partition (see {ref}`partition-names`).

You can find more information about all available partitions here. 
```

Please remove the two "see partitions' links and just make the last sentence I added a link to that page.

::::

::::{dropdown} 3. Is there a limit on the number of CPUs or GPUs that can be requested when trying to use the `lowpriority` partition?

> The maximum limit for number of CPUs, GPUs, job time etc. will be the same as that on the `short` and `gpu` partitions. Please check our partitions (see {ref}`partition-names`) page for the core and RAM limits on these partition. 

Reviewer 2: ```suggestion
The maximum limit for the number of CPUs and GPUs is the same as the `short` and `gpu` partitions as found on the  {ref}`partition-names` page. 
```

Please make "partitions page" the link.

::::

> ::::{dropdown} 4. I am an owner of a private partition, will this affect my group’s job wait time?

Reviewer 2: ```suggestion
::::{dropdown} 4. I am an owner of a private partition, will this affect my group’s wait time?
```

Reviewer: Addressed and pushed. Thank you!!

> As a PI who purchased your own equipment, jobs submitted by the members of your private partition to your own resources always have the highest priority. When submitting jobs to your private partition, any low priority jobs currently running on your servers will be killed and sent back to the queue, introducing a slight delay (~30s) on the first job submitted. This new service is designed to favor partition owners heavily.

Reviewer 2: ```suggestion
Jobs submitted by the members of your private partition to your privately owned resources always have the highest priority. When submitting jobs to your private partition, any low priority jobs currently running on your servers will be killed and sent back to the queue, introducing a slight delay (~30s) on the first job submitted. This new service is designed to favor partition owners heavily.
```

Reviewer: Addressed and pushed. Thank you!!

::::

::::{dropdown} 5. Will the `lowpriority` partition affect the limits on my own server(s)?

> The `lowpriority` partition will not affect current partition definitions — it is just an additional partition. As is true now, PIs will be able to request their private partitions be configured according to their research requirements. 

Reviewer 2: ```suggestion
The `lowpriority` partition will not affect current partition definitions — it is an additional partition. As is true now, PIs will be able to request their private partitions be configured according to their research requirements. 
```

Reviewer: Addressed and pushed. Thank you!!

::::

::::{dropdown} 6. Will the software that my group uses exclusively on our server(s) be available on other servers?

Yes, all software used in the private partitions is available on all servers. Software that is restricted to a particular group’s license will remain restricted to members of that group, but they will be able to use that software on any server. 
::::

::::{dropdown} 7. What hardware will be part of the `lowpriority` partition?
All PI-owned hardware purchased on or after 2019 will be part of the `lowpriority` partition.
::::

::::{dropdown} 8. Do you have a clear inventory of the various resources available through the `lowpriority` partition?

> Throughout the testing phase, RC will be updating its `low priority` documentation (see {ref}`low-priority`)to reflect the resources that are part of public and private partitions. Please check the documentation (see {ref}`low-priority`) regularly for latest information. See our CPUs ({ref}`hardware-overview`) and GPUs ({ref}`working-gpus`) pages.

Reviewer 2: This answer is really choppy and confusing. Is this question actually necessary, or can we just remove it completely? It just links to a bunch of pages we already link to in other areas and commits us to updating these pages frequently. If we do need to keep it we need to clean up the links.

Reviewer: It might be nice to keep just for the GPUs to be honest. The CPUs I don't think would be as important but some ML/DL workflows require certain memory requirements and it would be beneficial for them to know what is available in the partition. 

Maybe an easier way to query the system and see what is available would be better so we don't have to be committed to updating these pages frequently.

::::

:::{button-link} faq.html
:color: primary
:align: center
Back to the top
:::
(low-priority-preemption)=
## Preemption

::::{dropdown} 1. What does 'preemption' or 'preemptable job' mean?

> In SLURM terminology, preemption is a scheduling mechanism that involves stopping/suspending one or more low priority jobs, to accommodate a high priority job (i.e. job submitted by the partition owner’s group member). A low priority job that can get suspended in this manner is also known as a ‘preemptable’ job. Preemptable jobs can run on private hardware owned by other research groups, with the potential risk of being suspended before their wall time ends if the private partition’s group member submits a job and requests for resources.

Reviewer 2: ```suggestion
In Slurm terminology, preemption is a scheduling mechanism that involves stopping/suspending one or more low priority jobs, to accommodate a high priority job. Preemptable jobs can run on private hardware owned by other research groups, with the potential risk of being suspended before their wall time ends if the private partition’s group member submits a job and requests for resources.
```

Reviewer: Addressed and pushed. Thank you!!

::::

::::{dropdown} 2. What do I need to know about preemption and are there any drawbacks for my research group?

**For private partition owners:** 

> When low priority jobs are running on your hardware, you may experience a slight delay (~30s) when submitting your first job to your private partition. This is the time it will take the scheduler to kill/preempt low priority jobs running on your partition. This new service is designed to minimize that delay.

Reviewer 2: ```suggestion
When low priority jobs are running on your hardware, you may experience a slight delay (~30s) when submitting your first job to your private partition while the scheduler preempts the low priority jobs running on your partition. This new service is designed to minimize that delay.
```

Reviewer: Addressed and pushed. Thank you!!

**For users of the** `lowpriority` **partition:** 

> Submitting jobs to this partition will make your jobs preemptable. That means they may be stopped/suspended and re-queued at any time. We recommend that you use strategies such as checkpointing (see {ref}`checkpoint-jobs`) to take full advantage of the `lowpriority` partition. RC is working on detailed documentation (see {ref}`low-priority`) and training sessions that will be available later in the Fall 2023 semester, once the service is ready to be released.

Reviewer 2: ```suggestion
Submitting jobs to this partition will make your jobs preemptable, which means they may be stopped/suspended and re-queued at any time; {ref}`checkpoint-jobs`) can help you take full advantage of the {ref}`low-priority`.
```

::::

> ::::{dropdown} 3. How does re-queuing of preempted (terminated) low priority jobs work?

Reviewer 2: Maybe change this to something like, What happens if my job is preempted?

Reviewer: Would 'What happens if my job is preempted on low-priority?' work? Maybe being explicit on which partition might head off further questions?

> Preempted jobs are put back in the default partition queue (`short`) and scheduled normally.

Reviewer 2: ```suggestion
Preempted jobs are put into the default partition queue (`short`) and scheduled normally.
```

Reviewer: Addressed and pushed. Thank you!!

::::

::::{dropdown} 4. My servers are being used by others, how fast can we retrieve them?

> As a private partition owner you will always have the highest priority when accessing your own resources. When submitting jobs to your private partition, any low priority jobs currently running on your servers will be killed and sent back to the queue. You will retrieve your server(s) within approximately 30s.

Reviewer 2: ```suggestion
As a private partition owner you will always have the highest priority when accessing your own resources. When submitting jobs to your private partition, any low priority jobs currently running on your servers will be preempted and sent back to the default (`short`) queue. You will retrieve your server(s) within approximately 30s.
```

Reviewer: Addressed and pushed. Thank you!!

::::

::::{dropdown} 5. I already have access to one or more of the large, long, and/or multigpu partition(s), do I benefit from using the `lowpriority` partition?

> The goal of `lowpriority` partition is to double the resources available to Discovery users. Hence, specifying the `lowpriority` partition in your SLURM job header gives your job a higher chance of being allocated resources, even when your first choice of partition is unavailable. For e.g., when you specify `#SBATCH --partition=short,lowpriority`, your job can start running on the `lowpriority` partition even when `short` is unavailable. 

Reviewer 2: ```suggestion
The goal of `lowpriority` partition is to double the resources available to Discovery users. Specifying the `lowpriority` partition in your Slurm job header gives your job a higher chance of being allocated resources when your first choice of partition is unavailable. For example, when you specify `#SBATCH --partition=short,lowpriority`, your job can start running on the `lowpriority` partition when `short` is unavailable. 
```

Reviewer: Corrected and pushed. Thank you!!

::::

:::{button-link} faq.html
:color: primary
:align: center
Back to the top
:::

## RC Policy

::::{dropdown} 1. Can I opt-out?

> We would prefer if you did not, but can if it's really required due to constraints like funding stipulations. Opening up idle private resources that consume power and cooling makes the cluster more useful for everyone (including you). This is a common practice in HPC implemented by many other universities and national labs. Also, all private hardware purchases that were placed since 2019 came with the explicit understanding that they would be added to the `lowpriority` partition when it became available (see next question).

Reviewer 2: ```suggestion
Opening up idle private resources that consume power and cooling makes the cluster more useful for everyone (including you). This is a common practice in HPC implemented by many other universities and national labs. Also, all private hardware purchases that were placed since 2019 came with the explicit understanding that they would be added to the `lowpriority` partition when it became available (see next question).

If your private resources have constraints (i.e., funding stipulations), please reach out to the RC team and we will remove your resources from the low priority partition.
```

Reviewer: I made the change and pushed it. Thank you!!

::::

::::{dropdown} 2. What is the understanding between RC and private partition owners?
The following understanding applies to all individual private partition owners who purchased servers hosted at MGHPCC since 2019:

RC racks, installs and maintains privately owned servers; RC purchases the equipment needed to connect these servers to the infiniband network fabric; RC pays for power and cooling to operate the servers during their entire lifetime; In return, private partition owners make their servers available to the `lowpriority` partition.
::::

> ::::{dropdown} 3. Are there policies or other formal documents available?

Reviewer 2: We should remove this question until we have the MoU completed and distributed.

Reviewer: I just removed that question and updated the numbering in the section.

> ::::{dropdown} 3. Are there policies or other formal documents available?

Reviewer 2: ```suggestion
::::{dropdown} 
```

> RC is currently drafting a memorandum of understanding (MoU) that will be distributed to all private partitions owners later in the Fall semester 2023. This document will formalize the understanding that was stated to all private partitions owners at the time of purchase, since 2019. Moreover, RC is working on a detailed set of documentation, service level objectives and training sessions that will be available later in the Fall 2023 semester, once the service is ready to be released.

Reviewer 2: ```suggestion
```

::::

::::{dropdown} 4. I have resources I would like to make available to the rest of the research community, what can I do?

All hardware purchased on or after 2019 will be part of the `lowpriority` partition. Feel free to contact us if you purchased hardware before 2019 and would like to make it available to the `lowpriority` partition.
::::

::::{dropdown} 5. Why should I make my private resources available to the rest of the research community?

> By making your private resources available to others through the `lowpriority` partition, you will allow the Northeastern research community to use them when they are idle. This, in time, could double the resources available to NURC users and is a common practice in HPC clusters to optimize the use of idle private resources that consume power and cooling. 

Reviewer 2: ```suggestion
By making your private resources available to others through the `lowpriority` partition, you will allow the Northeastern research community to use them when they are idle. This, in time, could double the resources available to Northeastern researchers. This is a common practice in HPC clusters to optimize the use of idle private resources that consume power and cooling. 
```

::::

:::{button-link} faq.html
:color: primary
:align: center
Back to the top
:::

> ## Testing and Release

Reviewer 2: This whole section needs to be updated or removed.

Reviewer: This feels like it could be removed because I don't think it will be a frequently asked question. We can talk about it maybe in a (ll/B)AMA and it is an easy removal.

::::

::::{dropdown} 2. What is the timeline of the testing phase?
The `lowpriority` partition will be available for testing in Fall 2023. Only groups who opt in will have access to the `lowpriority` partition during the testing phase and only hardware owned by those groups will be part of the test partition.
::::

::::{dropdown} 3. How can I participate in the testing phase?
Let us know by sending an email to [rchelp@northeastern.edu](mailto:rchelp%40northeastern.edu) with the subject “lowpriority partition”. Members of your group will get first access to all newly opened resources for the duration of the testing phase. Only hardware owned by PIs who opt-in will be part of the testing phase.
::::

::::{dropdown} 4. When will the `lowpriority` partition be officially released?
Once testing, configuration, and documentation are complete, we will schedule an additional rollout of the `lowpriority` partition and provide access to all users of the HPC cluster. We anticipate this to be in **late 2022/early 2023**. 
::::

:::{button-link} faq.html
:color: primary
:align: center
Back to the top
:::

> ## Miscellaneous

Reviewer 2: I think we could just remove this whole section as well. We reference the LPP and Checkpointing pages throughout this entire FAQ.
